### PR TITLE
wormchain - contract deploy + test

### PR DESCRIPTION
**Pull request by justinschuldt**
This adds cosmwasm contract deploy for wormchain in tilt, for local development, and as a standalone script for CI testing.

Couple things here:
- addresses and such added to `devnet-consts.json`
- added some code to copy `devnet-consts.json` similar to how `.env` files are added to chain directories
- happy path tests for contract interacton
- had to add libwasmvm.x86_64.so to the wormchain docker image to get it build

### CODE REVIEW

**Comment by justinschuldt** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2149** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/2149" target="_blank">Graphite</a> 
    * **PR #2150** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/2150" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/2150?utm_source=stack-comment).

File: scripts/distribute-devnet-consts.sh
Comment (jynnantonix) [inline scripts/distribute-devnet-consts.sh:10]: style nit: use an array
```suggestion
paths=(
    ./terra2/tools/
    ./wormchain/contracts/tools/
)

for dest in "${paths[@]}"; do
```

File: wormchain/Makefile
Comment (jynnantonix) [inline wormchain/Makefile:72]: In the past I've had a problem with this kind of touch command because if the `node_modules` directory doesn't already exist then this will put a file there, which will cause the subsequent `npm ci` command to fail because `node_modules` isn't a directory.

File: wormchain/Makefile
Comment (jynnantonix) [inline wormchain/Makefile:82]: nit: use `$(MAKE)`

File: wormchain/build/config/genesis.json
Comment (jynnantonix) [inline wormchain/build/config/genesis.json:65]: Is this going to conflict with the worm tokens that are already minted elsewhere?

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (jynnantonix) [inline wormchain/contracts/tools/deploy_wormchain.ts:35]: Now that #2121 is merged, we don't actually need to deploy any of these.  We only need to deploy the accounting contract itself.

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (jynnantonix) [inline wormchain/contracts/tools/deploy_wormchain.ts:290]: If you want to wait until #2131 is merged, all this can be dropped as well.

File: wormchain/build/config/genesis.json
Comment (conorpp) [reply wormchain/build/config/genesis.json:65]: It might, but this isn't the same as the mainnet genesis file which has removed uworm.  I don't see an issue with leaving it in devnet or just removing it as well.

**Review by jynnantonix — APPROVED** [review]
There are still a couple of unaddressed comments from the previous review but otherwise looks good.

File: scripts/distribute-devnet-consts.sh
Comment (justinschuldt) [reply scripts/distribute-devnet-consts.sh]: good suggestion, fixed

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts:35]: I decided to leave the core contract and token bridge deploy since it is already written, and we'll likely want it in devnet in the future.

Comment (jynnantonix) [reply wormchain/contracts/tools/deploy_wormchain.ts:35]: I don't think there is any situation where we will want to deploy the core contract on wormchain because wormchain has the native core module, which does everything the core contract does.

**Review by a5-pickle — APPROVED** [review]

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (nik-suri) [reply wormchain/contracts/tools/deploy_wormchain.ts:35]: +1, no need to deploy any of the bridge contracts for now. I think the only one we need on wormchain is the accounting contract.

File: wormchain/contracts/tools/.npmrc
Comment (nik-suri) [inline wormchain/contracts/tools/.npmrc:1]: can we add a `.nvmrc` file as well to pin the node version?

**Review by conorpp — APPROVED** [review]

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts:35]: Sure, I'll update the deploy script to only deploy the accountant contract.

Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts]: done, resolving

File: wormchain/contracts/tools/.npmrc
Comment (justinschuldt) [reply wormchain/contracts/tools/.npmrc:1]: This falls under the same node version requirements as the rest of the monorepo. I don't think it makes sense to add one here but no where else.

**Comment by justinschuldt** [comment]
Got this rebased, deploy + tests are running against 2 wormchain nodes in CI. Please review and stamp when you can!

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (jynnantonix) [inline wormchain/contracts/tools/deploy_wormchain.ts:106]: I'm surprised this works.  I thought we changed things so that contracts could only be uploaded via a governance action in #1674.

@conorpp thoughts?

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (jynnantonix) [inline wormchain/contracts/tools/deploy_wormchain.ts:116]: Same here.  I thought instantiation could only be done through governance.

File: wormchain/contracts/tools/package.json
Comment (jynnantonix) [inline wormchain/contracts/tools/package.json:13]: Why ISC?  AFAIU the rest of the repo is Apache-2

File: wormchain/contracts/tools/test_wormchain.ts
Comment (jynnantonix) [inline wormchain/contracts/tools/test_wormchain.ts:169]: This will no longer be an error once #2241 merges.  Instead we'll return a response with an error for the observation status but the tx itself will succeed.

**Review by jynnantonix — APPROVED** [review]
I'm fine with merging but I do have some open questions.

File: wormchain/devnet/client/config/config.toml
Comment (hendrikhofstadt) [inline wormchain/devnet/client/config/config.toml:252]: why is pex disabled?

File: wormchain/devnet/base/config/config.toml
Comment (hendrikhofstadt) [inline wormchain/devnet/base/config/config.toml:252]: same here

**Review by evan-gray — DISMISSED** [review]
Can you explain why this differs significantly from the deployment in https://github.com/wormhole-foundation/wormhole/pull/2117/files#diff-79cdcc88d86ff8d9eb545c14eaa47f60d627c1c0935f5e3fb989efcf78c7b015 particularly, as @jynnantonix pointed out, I’m not needing the guardian signature to deploy and initialize? Also why were all these wormchain toml files and etc changes required?  Also, why were the wormchain token bridge registrations removed?

**Review by hendrikhofstadt — APPROVED** [review]
Why are we no longer registering wormchain? At some point we will want to test the token bridge on wormchain, is the plan to just readd the registrations then?

**Review by bruce-riley — DISMISSED** [review]

File: wormchain/devnet/client/config/config.toml
Comment (justinschuldt) [reply wormchain/devnet/client/config/config.toml:252]: If pex is enabled on non seed-nodes (in devnet, wormchain instances >= 1) they produce tons of errors as they try to find and connect to peers. We only need need pex on the seed node (`wormchain-0`), we don't need all nodes to be scanning the network for peers. Same reason/change as https://github.com/wormhole-foundation/wormhole/pull/2272.

File: wormchain/devnet/base/config/config.toml
Comment (justinschuldt) [reply wormchain/devnet/base/config/config.toml:252]: same as other, it's off by default, turned on in `wormchain/devnet/create-config.sh`.

File: wormchain/contracts/tools/test_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/test_wormchain.ts:169]: Good to know, thanks for the flag.

**Comment by justinschuldt** [comment]
> Can you explain why this differs significantly from the deployment in https://github.com/wormhole-foundation/wormhole/pull/2117/files#diff-79cdcc88d86ff8d9eb545c14eaa47f60d627c1c0935f5e3fb989efcf78c7b015 particularly, as @jynnantonix pointed out, I’m not needing the guardian signature to deploy and initialize? Also why were all these wormchain toml files and etc changes required? Also, why were the wormchain token bridge registrations removed?

@evan-gray 
- Differences from terra2 deploy - this does not deploy the wormhole core contract, just accountant, see the relevant discussion [here](https://github.com/wormhole-foundation/wormhole/pull/2150#discussion_r1052999832).
- Contract deploy permissions - this is a vanilla cosmwasm deploy/instantiate/invoke. Wormchain deploy through governance is TBD (currently not required, as this proves). We will decide what we want and address in the future.
- wormchain toml tiles changes - the files that were checked in had the default comments stripped out and some config items deleted, which made it harder to grok than it should be. This adds description comments so it's easier to understand what the options do and why things work like they do.
- No Token Bridge - token bridge currently requires the core contract, which decidedly should not be deployed to wormchain ([discussion](https://github.com/wormhole-foundation/wormhole/pull/2150#discussion_r1052999832)).

**Comment by justinschuldt** [comment]
> Why are we no longer registering wormchain? At some point we will want to test the token bridge on wormchain, is the plan to just readd the registrations then?

@hendrikhofstadt See the discussion [here](https://github.com/wormhole-foundation/wormhole/pull/2150#discussion_r1052999832).
We're not deploying the wormhole core contract, because that functionality is replaced by the native wormchain module. The token bridge contract requires wormhole core contract be deployed and initialized, so it can read the guardian set, etc.
We will need to modify the token bridge contract for wormchain to read the guardian set from wormchain's state, rather than from wormhole core's state. I made an issue to track this work https://github.com/wormhole-foundation/wormhole/issues/2275.

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts:106]: You are correct, the deploy permissions are not effective. Conor found the issue, we will implement permissioned deploy in the future, tracking with https://github.com/wormhole-foundation/wormhole/issues/2276.

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts:116]: Same as other, tracking with https://github.com/wormhole-foundation/wormhole/issues/2276.

File: wormchain/contracts/tools/package.json
Comment (justinschuldt) [reply wormchain/contracts/tools/package.json:13]: removing

File: wormchain/contracts/tools/test_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/test_wormchain.ts:169]: removing

File: wormchain/contracts/tools/deploy_wormchain.ts
Comment (justinschuldt) [reply wormchain/contracts/tools/deploy_wormchain.ts:33]: The intent is for `make contracts-deploy-setup` to be run before this deploy, to get the prereqs in place. `artifacts` will be copied to this path. I added a try/catch around this with a log reminding the user to run the `make` command. Appreciate the note.

**Comment by justinschuldt** [comment]
I addressed the nits and rebased, one more stamp plz @jynnantonix @hendrikhofstadt

**Review by jynnantonix — APPROVED** [review]

**Review by bruce-riley — DISMISSED** [review]

**Review by hendrikhofstadt — CHANGES_REQUESTED** [review]
I understand that the token bridge needs to interact with the native module, however the code for registering the token bridge will stay the same (as it's still going to be a wasm module). Manually removing the code here means that whoever is going to make said modification to the token bridge will have to reintroduce the code removed here.

File: wormchain/build/config/client.toml
Comment (hendrikhofstadt) [inline wormchain/build/config/client.toml:14]: this is a listen address but there should be a target ip here (0.0.0.0 != 127.0.0.1)

File: wormchain/build/config/config.toml
Comment (hendrikhofstadt) [inline wormchain/build/config/config.toml:96]: this may break things

Comment (justinschuldt) [reply wormchain/build/config/config.toml:96]: there are no websites that interact with the local development build.

File: wormchain/build/config/client.toml
Comment (justinschuldt) [reply wormchain/build/config/client.toml:14]: `0.0.0.0` is the cosmwasm default. the `/build/config/` is for local development, where the devnet config is in `/devnet/`

**Comment by justinschuldt** [comment]
> I understand that the token bridge needs to interact with the native module, however the code for registering the token bridge will stay the same (as it's still going to be a wasm module). Manually removing the code here means that whoever is going to make said modification to the token bridge will have to reintroduce the code removed here.

@hendrikhofstadt I'm fine with including token bridge, or not, I'm just trying to execute what the group wants. I agree all the (now removed) deploy & test code for token bridge is useful and will be needed in the future. However, the reviewers in this thread seem to agree the core contract should not be deployed, and the token bridge deploy currently requires the core contract.

We need contracts deployed to devnet for Global Accountant - this is a blocker. We do not currently need Token Bridge deployed for any reason, so I don't think we should hold this PR back for a nice-to-have.

I'm going to make a graphite branch stacked on this one, and add the token bridge deploy code to that, and we can discuss/decide what to do for token bridge there.

**Comment by justinschuldt** [comment]
@hendrikhofstadt made this PR to discuss token bridge contract options: https://github.com/wormhole-foundation/wormhole/pull/2282

I'm dismissing the change request and merging this now.

**Review by hendrikhofstadt — DISMISSED** [review]

**Review by bruce-riley — APPROVED** [review]

**Review by conorpp — DISMISSED** [review]

**Comment by justinschuldt** [comment]
I'm going to break this into a stack of smaller PRs so it is more manageable to get approvers and merge.

No need to review, you'll get pinged in the new PRs

**Comment by evan-gray** [comment]
Closing in favor of the stack starting with #2296

**Pull request closed without merging**
